crates/farming/src/lib.rs

```diff
@@ -0,0 +1,352 @@
+            amount: BalanceOf<T>,
+        },
+        RewardClaimed {
+            account_id: AccountIdOf<T>,
+            pool_currency_id: CurrencyIdOf<T>,
+            reward_currency_id: CurrencyIdOf<T>,
+            amount: BalanceOf<T>,
+        },
+    }
+
+    #[pallet::error]
+    pub enum Error<T> {
+        ScheduleNotFound,
+    }
+
+    #[pallet::hooks]
+    impl<T: Config> Hooks<T::BlockNumber> for Pallet<T> {
+        fn on_initialize(now: T::BlockNumber) -> Weight {
+            if now % T::RewardPeriod::get() == Zero::zero() {
+                let mut count: u32 = 0;
+                // collect first to avoid modifying in-place
+                let schedules = RewardSchedules::<T>::iter().collect::<Vec<_>>();
+                for (pool_currency_id, reward_currency_id, mut reward_schedule) in schedules.into_iter() {
+                    if let Some(amount) = reward_schedule.take() {
+                        if let Ok(_) = Self::try_distribute_reward(pool_currency_id, reward_currency_id, amount) {
+                            // only update the schedule if we could distribute the reward
+                            RewardSchedules::<T>::insert(pool_currency_id, reward_currency_id, reward_schedule);
+                            count.saturating_inc();
+                            Self::deposit_event(Event::RewardDistributed {
+                                pool_currency_id,
+                                reward_currency_id,
+                                amount,
+                            });
+                        }
+                    } else {
+                        // period count is zero
+                        RewardSchedules::<T>::remove(pool_currency_id, reward_currency_id);
+                        // TODO: sweep leftover rewards?
+                    }
+                }
+                T::WeightInfo::on_initialize(count)
+            } else {
+                Weight::zero()
+            }
+        }
+    }
+
+    #[pallet::storage]
+    #[pallet::getter(fn reward_schedules)]
+    pub type RewardSchedules<T: Config> = StorageDoubleMap<
+        _,
+        Blake2_128Concat,
+        CurrencyIdOf<T>, // lp token
+        Blake2_128Concat,
+        CurrencyIdOf<T>, // reward currency
+        RewardScheduleOf<T>,
+        ValueQuery,
+    >;
+
+    #[pallet::pallet]
+    pub struct Pallet<T>(_);
+
+    // The pallet's dispatchable functions.
+    #[pallet::call]
+    impl<T: Config> Pallet<T> {
+        /// Create or overwrite the reward schedule, if a reward schedule
+        /// already exists for the rewards currency the duration is added
+        /// to the existing duration and the rewards per period are modified
+        /// s.t. that the total (old remaining + new) rewards are distributed
+        /// over the new total duration
+        #[pallet::weight(T::WeightInfo::update_reward_schedule())]
+        #[transactional]
+        pub fn update_reward_schedule(
```

Comment:
I'm wondering how we would achieve the following: We would like to set non-linear reward schedules easily. For example, we might want to distribute incentives over 3 months in a 60:30:10 fashion. Liquidity in pools tends to be sticky, so initial high rewards tends to attract high initial liquidity whereas over time the additional rewards can be reduced.

With the current API, the best approach would be to use the scheduler pallet to make the governance call to initialize the rewards and then update them twice. I'm wondering if we could configure such a thing with a simple batch on the farming pallet directly instead.

Long-term, others might want to use our AMM as well and farming is such a common feature that it'll likely be worth to make this API as simple as possible.

Reply:
I decided against unnecessarily complicating the farming pallet for now since as you noted it is possible to simply schedule the calls which we can batch.

Reply:
Ok, that makes sense. Let's leave it as is then!